addons/binding/org.openhab.binding.freebox/src/main/java/org/openhab/binding/freebox/discovery/FreeboxDiscoveryService.java

```diff
@@ -0,0 +1,100 @@
+/**
+ * Copyright (c) 2014 openHAB UG (haftungsbeschraenkt) and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.freebox.discovery;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+import org.matmaul.freeboxos.FreeboxException;
+import org.matmaul.freeboxos.FreeboxOsClient;
+import org.matmaul.freeboxos.login.Authorize;
+import org.matmaul.freeboxos.login.LoginManager;
+import org.osgi.framework.Bundle;
+import org.osgi.framework.FrameworkUtil;
+import org.eclipse.smarthome.config.discovery.AbstractDiscoveryService;
+import org.eclipse.smarthome.config.discovery.DiscoveryResult;
+import org.eclipse.smarthome.config.discovery.DiscoveryResultBuilder;
+import org.eclipse.smarthome.core.thing.ThingTypeUID;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static org.openhab.binding.freebox.FreeboxBindingConstants.*;
+
+/**
+ * The {@link FreeboxDiscoveryService} is responsible for discovering a Freebox Thing
+ * on the local network
+ * 
+ * @author Gaël L'hopital - Initial contribution
+ */
+
+public class FreeboxDiscoveryService extends AbstractDiscoveryService {
+
+	private static final Logger logger = LoggerFactory.getLogger(FreeboxDiscoveryService.class);
+
+	public FreeboxDiscoveryService() {
+		super(SUPPORTED_THING_TYPES_UIDS,10);
+	}
+
+	public Set<ThingTypeUID> getSupportedThingTypes() {
+		return SUPPORTED_THING_TYPES_UIDS;
+	}
+	
+	@Override
+	protected void startScan() {
+		discoverFreebox();
+	}
+
+
+	/**
+	 * Discover the Freebox, get its appToken then submits the result
+	 */
+	private void discoverFreebox(){
+		Bundle bundle = FrameworkUtil.getBundle(getClass());			
+		
+		FreeboxOsClient fbClient = new FreeboxOsClient(
+					bundle.getSymbolicName() /*  org.openhab.binding.freebox */
+					);
+		
+		LoginManager loginManager = fbClient.getLoginManager();
+		
+		try {
+			Authorize authorize = loginManager.newAuthorize(
```

Comment:
I'm a bit reluctant, because once it is merged it will serve as an example for others how to implement bindings - and we might end up with many such solutions...

Think about it this way: You anyhow have to cover the case that your token might become invalidated somewhen. So you need a way to reauthenticate when the thing already exists and the token is not accepted. So why not implement this logic right now? And if it is there, you can easily remove the authentication from the discovery, because this can be done when the thing has been created.